pageserver/src/page_service.rs

```diff
@@ -899,165 +995,383 @@ impl PageServerHandler {
+                            let mut guard = pending_batch.lock().unwrap();
+                            match &mut *guard {
+                                BatchState::Building(batch) => {
+                                    *guard = BatchState::ReadMessagesEnded(batch.take());
+                                }
+                                BatchState::ReadMessagesEnded(_) => {
+                                    unreachable!("we exit the first time")
+                                }
+                                BatchState::ExecutorEnded => {
+                                    debug!("observing executor ended when reading upstream");
+                                }
                             }
-                            PageStreamError::Reconnect(reason) => {
-                                span.in_scope(|| info!("handler requested reconnect: {reason}"));
-                                return Err(QueryError::Reconnect);
+                        });
+                        break;
+                    };
+                    // don't read new requests before this one has been processed
+                    let mut req = Some(req);
+                    loop {
+                        let mut wait_notified = None;
+                        enum Outcome {
+                            Batched,
+                            CannotBatchNeedWaitForExecutor,
+                            ExecutorEndObserved,
+                            Undefined,
+                        }
+                        let mut outcome = Outcome::Undefined;
+                        batch_tx.send_if_modified(|pending_batch| {
+                            let mut guard = pending_batch.lock().unwrap();
+                            let building = match &mut *guard {
+                                BatchState::Building(building) => building,
+                                BatchState::ReadMessagesEnded(_) => {
+                                    unreachable!("we would have bailed earlier")
+                                }
+                                BatchState::ExecutorEnded => {
+                                    debug!("observing executor ended when trying to batch");
+                                    outcome = Outcome::ExecutorEndObserved;
+                                    return false;
+                                }
+                            };
+                            match Self::pagestream_do_batch(
+                                max_batch_size,
+                                building,
+                                req.take().unwrap(),
+                            ) {
+                                Some(req_was_not_batched) => {
+                                    outcome = Outcome::CannotBatchNeedWaitForExecutor;
+                                    req.replace(req_was_not_batched);
+                                    wait_notified = Some(notify_batcher.notified());
```

Comment:
Can we wrap `wait_notified` inside the `Outcome::CannotBatchNeedWaitForExecutor` variant?